File: tools/build_claim_template.mjs
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { execFileSync } from "node:child_process";


const EXPECTED_SHA256 = "0ad72f03131b95f322af320960a847c63121dec8a6f0457438f9a5031f9b1432";


function fail(message) {
  throw new Error(message);
}


function replaceRequired(source, search, replacement, expectedCount = 1) {
  const parts = source.split(search);
  const actualCount = parts.length - 1;
  if (actualCount !== expectedCount) {
    fail(`Expected ${expectedCount} occurrence(s) of ${JSON.stringify(search)}, found ${actualCount}`);
  }
  return parts.join(replacement);
}


function replaceSequential(source, search, replacements) {
  let result = source;
  for (const replacement of replacements) {
    const index = result.indexOf(search);
    if (index < 0) {
      fail(`Missing sequential placeholder ${JSON.stringify(search)}`);
    }
    result = result.slice(0, index) + replacement + result.slice(index + search.length);
  }
  if (result.includes(search)) {
    fail(`Unexpected extra placeholder ${JSON.stringify(search)}`);
  }
  return result;
}


function visibleParagraphText(paragraphXml) {
  return paragraphXml
    .replace(/<[^>]+>/g, "")
    .replaceAll("&amp;", "&")
    .replaceAll("&lt;", "<")
    .replaceAll("&gt;", ">");
}


function removeParagraphs(documentXml, textFragments) {
  const found = new Map(textFragments.map((fragment) => [fragment, 0]));
  const result = documentXml.replace(/<w:p\b[\s\S]*?<\/w:p>/g, (paragraphXml) => {
    const text = visibleParagraphText(paragraphXml);
    for (const fragment of textFragments) {
      if (text.includes(fragment)) {
        found.set(fragment, found.get(fragment) + 1);
        return "";
      }
    }
    return paragraphXml;
  });
  for (const [fragment, count] of found) {
    if (count !== 1) {
      fail(`Expected one paragraph containing ${JSON.stringify(fragment)}, found ${count}`);
    }
  }
  return result;
}


function replaceParagraphText(documentXml, sourceText, replacementText) {
  let count = 0;
  const result = documentXml.replace(/<w:p\b[\s\S]*?<\/w:p>/g, (paragraphXml) => {
    if (visibleParagraphText(paragraphXml) !== sourceText) {
      return paragraphXml;
    }
    count += 1;
    const openingTag = paragraphXml.match(/^<w:p\b[^>]*>/)?.[0];
    const paragraphProperties = paragraphXml.match(/<w:pPr>[\s\S]*?<\/w:pPr>/)?.[0] || "";
    const runProperties = paragraphXml.match(/<w:rPr>[\s\S]*?<\/w:rPr>/)?.[0] || "";
    if (!openingTag) {
      fail(`Could not preserve paragraph wrapper for ${JSON.stringify(sourceText)}`);
    }
    return `${openingTag}${paragraphProperties}<w:r>${runProperties}<w:t>${replacementText}</w:t></w:r></w:p>`;
  });
  if (count !== 1) {
    fail(`Expected one paragraph equal to ${JSON.stringify(sourceText)}, found ${count}`);
  }
  return result;
}


function moveExplicitPageBreakToHeading(documentXml, headingText) {
  let breakCount = 0;
  let headingCount = 0;
  let result = documentXml.replace(/<w:p\b[\s\S]*?<\/w:p>/g, (paragraphXml) => {
    if (paragraphXml.includes('<w:br w:type="page"/>')) {
      breakCount += 1;
      return "";
    }
    if (visibleParagraphText(paragraphXml) !== headingText) {
      return paragraphXml;
    }
    headingCount += 1;
    if (!paragraphXml.includes("</w:pPr>")) {
      fail(`Heading ${JSON.stringify(headingText)} has no paragraph properties`);
    }
    return paragraphXml.replace("</w:pPr>", "<w:pageBreakBefore/></w:pPr>");
  });
  if (breakCount !== 1 || headingCount !== 1) {
    fail(`Expected one page break and one heading ${JSON.stringify(headingText)}; found ${breakCount}/${headingCount}`);
  }
  return result;
}


function main() {
  const [sourceArgument, outputArgument] = process.argv.slice(2);
  if (!sourceArgument || !outputArgument) {
    fail("Usage: node tools/build_claim_template.mjs SOURCE.docx OUTPUT.docx");
  }

  const sourcePath = path.resolve(sourceArgument);
  const outputPath = path.resolve(outputArgument);
  const sourceBytes = fs.readFileSync(sourcePath);
  const sha256 = crypto.createHash("sha256").update(sourceBytes).digest("hex");
  if (sha256 !== EXPECTED_SHA256) {
    fail(`Reference DOCX checksum mismatch: expected ${EXPECTED_SHA256}, got ${sha256}`);
  }

  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  const workDir = fs.mkdtempSync(path.join(os.tmpdir(), "ecopark-claim-template-"));
  try {
    execFileSync("unzip", ["-q", sourcePath, "-d", workDir]);
    const documentPath = path.join(workDir, "word", "document.xml");
    const settingsPath = path.join(workDir, "word", "settings.xml");
    let xml = fs.readFileSync(documentPath, "utf8");

    xml = removeParagraphs(xml, [
      "Приложения",
      "Расчет задолженности по состоянию",
      "Копия или выписка из документа",
      "Копии счетов, уведомлений",
      "[Иные документы при необходимости].",
    ]);
    xml = moveExplicitPageBreakToHeading(xml, "Требования ТСН");

    xml = replaceRequired(xml, "Адрес: [юридический / почтовый адрес ТСН]", "Адрес: {{TSN_ADDRESS}}");
    xml = replaceRequired(
      xml,
      "Телефон: [номер] | Электронная почта: [адрес]",
      "Телефон: {{TSN_PHONE}} | Электронная почта: {{TSN_EMAIL}}",
    );
    xml = replaceRequired(
      xml,
      "Собственнику земельного участка № [номер участка]",
      "Собственнику земельного участка № {{PLOT_NUMBER}}",
    );
    xml = replaceRequired(
      xml,
      "[Фамилия Имя Отчество / наименование собственника]",
      "{{OWNER_NAME}}",
    );
    xml = replaceRequired(
      xml,
      "Адрес участка: [адрес / кадастровый номер]",
      "Адрес участка: {{PLOT_ADDRESS}}",
    );
    xml = replaceRequired(xml, "Исх. № [номер]", "Исх. № {{CLAIM_NUMBER}}");
    xml = replaceRequired(xml, "от [дата] 2026 г.", "от {{CLAIM_DATE}}");
    xml = replaceRequired(xml, "[кадастровый номер]", "{{CADASTRAL_NUMBER}}");
    xml = replaceRequired(xml, "[адрес участка]", "{{PLOT_ADDRESS}}");
    xml = replaceParagraphText(
      xml,
      "Начисления произведены на основании решения общего собрания.",
      "Начисления произведены на основании {{CHARGE_BASIS}}.",
    );
    xml = replaceRequired(
      xml,
      "Сведения для сверки приведены ниже; подробный расчет является приложением к претензии.",
      "По данным бухгалтерского учета ТСН у Вас образовалась задолженность по лицевому счету {{ACCOUNT}} в указанном ниже размере.",
    );
    xml = replaceRequired(
      xml,
      "[дата, по состоянию на которую определен долг]",
      "{{CALCULATION_DATE}}",
    );
    xml = replaceRequired(xml, "[с __.__.____ по __.__.____]", "{{DEBT_PERIOD}}");
    xml = replaceRequired(xml, "[сумма] руб. [коп.] коп.", "{{PRINCIPAL_AMOUNT}}");
    xml = replaceRequired(
      xml,
      "[сумма либо «не начислены»; указать основание]",
      "{{PENALTY_TEXT}}",
    );
    xml = replaceRequired(
      xml,
      "[итоговая сумма] руб. [коп.] коп.",
      "{{TOTAL_AMOUNT}}",
    );
    xml = replaceParagraphText(
      xml,
      "В течение 10 календарных дней со дня получения настоящей претензии (если иной срок не установлен применимым законом или договором) погасить задолженность в размере [итоговая сумма] руб. [коп.] коп.",
      "В течение 10 календарных дней со дня получения настоящей претензии (если иной срок не установлен применимым законом или договором) погасить задолженность в размере {{TOTAL_AMOUNT}}",
    );
    xml = replaceRequired(xml, "[адрес]", "{{TSN_EMAIL}}");
    xml = replaceSequential(xml, "[номер]", [
      "{{PLOT_NUMBER}}",
      "{{CLAIM_NUMBER}}",
      "{{PLOT_NUMBER}}",
    ]);
    xml = replaceRequired(xml, "[дата]", "{{CLAIM_DATE}}");
    xml = replaceRequired(xml, "Претензию и приложения получил(а): ", "Претензию получил(а): ");
    xml = replaceRequired(xml, "[Ф. И. О. получателя]", "{{OWNER_NAME}}");

    const originalPlaceholders = xml.match(/\[[^\]]+\]/g) || [];
    if (originalPlaceholders.length) {
      fail(`Unconverted bracket placeholders: ${originalPlaceholders.join(", ")}`);
    }
    fs.writeFileSync(documentPath, xml, "utf8");

    let settingsXml = fs.readFileSync(settingsPath, "utf8");
    if (!settingsXml.includes("<w:updateFields")) {
      settingsXml = settingsXml.replace(
        "</w:settings>",
        '<w:updateFields w:val="true"/></w:settings>',
      );
    }
    fs.writeFileSync(settingsPath, settingsXml, "utf8");

    fs.rmSync(outputPath, { force: true });
    execFileSync("zip", ["-q", "-X", "-D", "-r", outputPath, "."], { cwd: workDir });
  } finally {
    fs.rmSync(workDir, { recursive: true, force: true });
  }
}


main();

File: tools/build_claim_template.py
#!/usr/bin/env python3
"""Build the application DOCX template from the approved Word reference."""

from hashlib import sha256
from html import unescape
from pathlib import Path
import re
import sys
from xml.sax.saxutils import escape
from zipfile import ZIP_DEFLATED, ZipFile


EXPECTED_SHA256 = "13024d74add6e7cea8c011a7ea79e4b4fef6e16dd2c25b953bbf29f08d19ca46"

CLAIM_TOKENS = (
    "TSN_ADDRESS",
    "TSN_PHONE",
    "TSN_EMAIL",
    "PLOT_NUMBER",
    "OWNER_NAME",
    "PLOT_ADDRESS",
    "CADASTRAL_NUMBER",
    "CLAIM_NUMBER",
    "CLAIM_DATE",
    "CHARGE_BASIS",
    "ACCOUNT",
    "CALCULATION_DATE",
    "DEBT_PERIOD",
    "PRINCIPAL_AMOUNT",
    "PENALTY_TEXT",
    "TOTAL_AMOUNT",
)

PARAGRAPH_PATTERN = re.compile(r"<w:p\b[\s\S]*?</w:p>")
ROW_PATTERN = re.compile(r"<w:tr\b[\s\S]*?</w:tr>")
CELL_PATTERN = re.compile(r"<w:tc\b[\s\S]*?</w:tc>")


def fail(message):
    raise RuntimeError(message)


def visible_text(xml):
    text_parts = re.findall(r"<w:t\b[^>]*>([\s\S]*?)</w:t>", xml)
    return unescape("".join(text_parts))


def replacement_paragraph(paragraph_xml, replacement):
    opening_tag = re.match(r"<w:p\b[^>]*>", paragraph_xml)
    if opening_tag is None:
        fail(f"Could not preserve paragraph wrapper for {visible_text(paragraph_xml)!r}")
    paragraph_properties = re.search(r"<w:pPr>[\s\S]*?</w:pPr>", paragraph_xml)
    run_properties = re.search(r"<w:rPr>[\s\S]*?</w:rPr>", paragraph_xml)
    ppr = paragraph_properties.group(0) if paragraph_properties else ""
    rpr = run_properties.group(0) if run_properties else ""
    value = escape(replacement)
    return (
        f'{opening_tag.group(0)}{ppr}<w:r>{rpr}'
        f'<w:t xml:space="preserve">{value}</w:t></w:r></w:p>'
    )


def replace_paragraph(document_xml, source_text, replacement):
    count = 0

    def callback(match):
        nonlocal count
        paragraph_xml = match.group(0)
        if visible_text(paragraph_xml).strip() != source_text.strip():
            return paragraph_xml
        count += 1
        return replacement_paragraph(paragraph_xml, replacement)

    result = PARAGRAPH_PATTERN.sub(callback, document_xml)
    if count != 1:
        fail(f"Expected one paragraph {source_text!r}, found {count}")
    return result


def remove_paragraph(document_xml, source_text):
    count = 0

    def callback(match):
        nonlocal count
        paragraph_xml = match.group(0)
        if visible_text(paragraph_xml).strip() != source_text.strip():
            return paragraph_xml
        count += 1
        return ""

    result = PARAGRAPH_PATTERN.sub(callback, document_xml)
    if count != 1:
        fail(f"Expected one removable paragraph {source_text!r}, found {count}")
    return result


def replace_table_value(document_xml, label, replacement):
    count = 0

    def row_callback(row_match):
        nonlocal count
        row_xml = row_match.group(0)
        cells = list(CELL_PATTERN.finditer(row_xml))
        if len(cells) < 2 or visible_text(cells[0].group(0)).strip() != label:
            return row_xml
        value_cell = cells[1]
        cell_xml = value_cell.group(0)
        paragraphs = list(PARAGRAPH_PATTERN.finditer(cell_xml))
        if not paragraphs:
            fail(f"Table row {label!r} has no value paragraph")
        paragraph = paragraphs[0]
        updated_cell = (
            cell_xml[: paragraph.start()]
            + replacement_paragraph(paragraph.group(0), replacement)
            + cell_xml[paragraph.end() :]
        )
        count += 1
        return row_xml[: value_cell.start()] + updated_cell + row_xml[value_cell.end() :]

    result = ROW_PATTERN.sub(row_callback, document_xml)
    if count != 1:
        fail(f"Expected one table row {label!r}, found {count}")
    return result


def assert_remaining_blanks_are_manual(document_xml):
    unexpected = []
    for paragraph in PARAGRAPH_PATTERN.findall(document_xml):
        text = visible_text(paragraph).strip()
        if "___" not in text:
            continue
        if text.startswith("___________________ / Дорогин"):
            continue
        if set(text) <= {"_"}:
            continue
        unexpected.append(text)
    if unexpected:
        fail(f"Unconverted fill-in blanks: {unexpected}")


def build(source_path, output_path):
    source_bytes = source_path.read_bytes()
    actual_sha256 = sha256(source_bytes).hexdigest()
    if actual_sha256 != EXPECTED_SHA256:
        fail(
            "Reference DOCX checksum mismatch: "
            f"expected {EXPECTED_SHA256}, got {actual_sha256}"
        )

    output_path.parent.mkdir(parents=True, exist_ok=True)
    with ZipFile(source_path, "r") as source, ZipFile(output_path, "w", ZIP_DEFLATED) as target:
        for info in source.infolist():
            payload = source.read(info.filename)
            if info.filename == "word/document.xml":
                xml = payload.decode("utf-8")
                replacements = (
                    (
                        "Адрес: 633204, Новосибирская область, г. Искитим,",
                        "Адрес: {{TSN_ADDRESS}}",
                    ),
                    (
                        "Телефон: +7 961 846-52-56 | Электронная почта: tsn-ecopark@mail.ru",
                        "Телефон: {{TSN_PHONE}} | Электронная почта: {{TSN_EMAIL}}",
                    ),
                    (
                        "Собственнику земельного участка № ___",
                        "Собственнику земельного участка № {{PLOT_NUMBER}}",
                    ),
                    ("__________________________________________", "{{OWNER_NAME}}"),
                    (
                        "Адрес участка: Новосибирская область, Искитимский муниципальный район, сельское поселение Быстровский сельсовет, с. Быстровка, микрорайон Экопарк, з/у № 10\"",
                        "Адрес участка: {{PLOT_ADDRESS}}",
                    ),
                    (
                        "Исх. № ____от ________________  г.",
                        "Исх. № {{CLAIM_NUMBER}}    от {{CLAIM_DATE}}",
                    ),
                    (
                        "Вы являетесь собственником земельного участка № _____, кадастровый номер ___________________, расположенного по адресу: Новосибирская область, Искитимский муниципальный район, сельское поселение Быстровский сельсовет, с. Быстровка, микрорайон Экопарк, з/у № ______ и членом ТСН «Микрорайон Экопарк».",
                        "Вы являетесь собственником земельного участка № {{PLOT_NUMBER}}, кадастровый номер {{CADASTRAL_NUMBER}}, расположенного по адресу: {{PLOT_ADDRESS}}, и членом ТСН «Микрорайон Экопарк».",
                    ),
                    (
                        "В соответствии с Уставом ТСН члены общества обязаны своевременно уплачивать членские и целевые взносы, размеры которых устанавливается решением Общего собрания членов ТСН.",
                        "В соответствии с Уставом ТСН члены товарищества обязаны своевременно уплачивать членские и целевые взносы, размеры которых устанавливаются решением общего собрания членов ТСН.",
                    ),
                    (
                        "Протоколом общего собрания от_________ установлены членские взносы в размере 285 руб./сотка.  По данным бухгалтерского учета ТСН у Вас образовалась задолженность по членским взносам в размере _______________:",
                        "Протоколом общего собрания от {{CHARGE_BASIS}} установлены членские взносы в размере 285 руб./сотка. По данным бухгалтерского учета ТСН по лицевому счету {{ACCOUNT}} у Вас образовалась задолженность по членским взносам в размере {{PRINCIPAL_AMOUNT}}:",
                    ),
                    (
                        "Обязанность по внесению взносов распространяется на всех членов товарищества. Целевые взносы вносятся членами товарищества на расчетный счет товарищества по решению общего собрания членов товарищества, определяющему их размер и срок внесения, в порядке, установленном уставом товарищества. Размер взносов определяется на основании приходно-расходной сметы товарищества и финансово-экономического обоснования, утвержденных общим собранием членов товарищества.Уставом товарищества может быть установлен порядок взимания и размер пеней в случае несвоевременной уплаты взносов. В случае неуплаты взносов и пеней товарищество вправе взыскать их в судебном порядке.",
                        "Обязанность по внесению взносов распространяется на всех членов товарищества. Целевые взносы вносятся членами товарищества на расчетный счет товарищества по решению общего собрания членов товарищества, определяющему их размер и срок внесения, в порядке, установленном уставом товарищества. Размер взносов определяется на основании приходно-расходной сметы товарищества и финансово-экономического обоснования, утвержденных общим собранием членов товарищества. Уставом товарищества может быть установлен порядок взимания и размер пеней в случае несвоевременной уплаты взносов. В случае неуплаты взносов и пеней товарищество вправе взыскать их в судебном порядке.",
                    ),
                    (
                        "В соответствии с п.7.5 Уставом ТСН член товарищества обязан нести бремя содержания своего земельного участка; вносить обязательные платежи и членские и целевые взносы, а также другие платежи, предусмотренные законодательством и Уставом, в размерах и в сроки, определяемые законодательством и Общим собранием членов товарищества; выполнять решения общего собрания членов товарищества и решения правления.",
                        "В соответствии с п. 7.5 Устава ТСН член товарищества обязан нести бремя содержания своего земельного участка; вносить обязательные платежи, членские и целевые взносы, а также другие платежи, предусмотренные законодательством и Уставом, в размерах и сроки, определяемые законодательством и общим собранием членов товарищества; выполнять решения общего собрания членов товарищества и решения правления.",
                    ),
                    (
                        "В связи с этим, задолженность по членскому взносу в размере _________ руб. подлежит взысканию с должника в пользу ТСН.",
                        "В связи с этим задолженность по членскому взносу в размере {{PRINCIPAL_AMOUNT}} подлежит взысканию с должника в пользу ТСН.",
                    ),
                    (
                        "Таким образом, взнос установлен решением общего собрания, которое не отменено, недействительным не признано, в связи с чем, подлежит уплате всеми членами товарищества.",
                        "Таким образом, взнос установлен решением общего собрания, которое не отменено и не признано недействительным, в связи с чем подлежит уплате всеми членами товарищества.",
                    ),
                    (
                        "На основании вышеизложенного, прошу незамедлительно рассмотреть настоящую досудебную претензию и произвести погашение образовавшейся задолженности в размере _______ рублей в течение 10 календарных дней со дня получения настоящей претензии (если иной срок не установлен применимым законом или договором. Претензия считается врученной любым способом, позволяющим установить ее получение должником.",
                        "На основании вышеизложенного прошу незамедлительно рассмотреть настоящую досудебную претензию и погасить образовавшуюся задолженность в размере {{TOTAL_AMOUNT}} в течение 10 календарных дней со дня получения настоящей претензии (если иной срок не установлен применимым законом или договором). Претензия считается врученной любым способом, позволяющим установить ее получение должником.",
                    ),
                    (
                        "В случае удовлетворения требований ТСН о погашении задолженности, прошу направить в ТСН подтверждение оплаты по адресу электронной почты tsn-ecopark@mail.ru либо передать копию платежного документа председателю ТСН.",
                        "В случае удовлетворения требований ТСН о погашении задолженности прошу направить в ТСН подтверждение оплаты по адресу электронной почты {{TSN_EMAIL}} либо передать копию платежного документа председателю ТСН.",
                    ),
                    (
                        "Назначение платежа: «Погашение задолженности по претензии № ___ от _____ _______ г., участок № _____». ",
                        "Назначение платежа: «Погашение задолженности по претензии № {{CLAIM_NUMBER}} от {{CLAIM_DATE}}, участок № {{PLOT_NUMBER}}».",
                    ),
                )
                for source_text, replacement in replacements:
                    xml = replace_paragraph(xml, source_text, replacement)

                xml = remove_paragraph(xml, "ул. Карбышева, д. 16")
                xml = replace_table_value(xml, "Расчетная дата", "{{CALCULATION_DATE}}")
                xml = replace_table_value(xml, "Период задолженности", "{{DEBT_PERIOD}}")
                xml = replace_table_value(xml, "Основной долг", "{{PRINCIPAL_AMOUNT}}")
                xml = replace_table_value(xml, "Пени / проценты", "{{PENALTY_TEXT}}")
                xml = replace_table_value(xml, "ИТОГО К ОПЛАТЕ", "{{TOTAL_AMOUNT}}")

                for token in CLAIM_TOKENS:
                    if "{{" + token + "}}" not in xml:
                        fail(f"Missing token {{{{{token}}}}} in generated template")
                assert_remaining_blanks_are_manual(xml)
                if "з/у № 10\"" in xml:
                    fail("Hard-coded plot 10 address remains in generated template")
                payload = xml.encode("utf-8")

            elif info.filename == "word/settings.xml":
                settings = payload.decode("utf-8")
                if "<w:updateFields" not in settings:
                    settings = settings.replace(
                        "</w:settings>",
                        '<w:updateFields w:val="true"/></w:settings>',
                    )
                payload = settings.encode("utf-8")

            target.writestr(info, payload)


def main():
    if len(sys.argv) != 3:
        fail("Usage: build_claim_template.py SOURCE.docx OUTPUT.docx")
    build(Path(sys.argv[1]).resolve(), Path(sys.argv[2]).resolve())


if __name__ == "__main__":
    main()
